src/Asset/TagRenderer.php

```diff
@@ -0,0 +1,76 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\WebpackEncoreBundle\Asset;
+
+use Symfony\Component\Asset\Packages;
+
+final class TagRenderer
+{
+    private $entrypointLookup;
+
+    private $manifestLookup;
+
+    private $packages;
+
+    public function __construct(EntrypointLookup $entrypointLookup, ManifestLookup $manifestLookup, Packages $packages = null)
+    {
+        $this->entrypointLookup = $entrypointLookup;
+        $this->manifestLookup = $manifestLookup;
+        $this->packages = $packages;
+    }
+
+    public function renderWebpackScriptTags(string $entryName, string $packageName = null): string
+    {
+        $scriptTags = [];
+        foreach ($this->entrypointLookup->getJavaScriptFiles($entryName) as $filename) {
+            $scriptTags[] = sprintf(
+                '<script src="%s"></script>',
+                $this->getAssetPath($filename, $packageName)
+            );
+        }
+
+        return implode('', $scriptTags);
+    }
+
+    public function renderWebpackLinkTags(string $entryName, string $packageName = null): string
+    {
+        $scriptTags = [];
+        foreach ($this->entrypointLookup->getCssFiles($entryName) as $filename) {
+            $scriptTags[] = sprintf(
+                '<link rel="stylesheet" href="%s" />',
+                $this->getAssetPath($filename, $packageName)
+            );
+        }
+
+        return implode('', $scriptTags);
+    }
+
+    private function getAssetPath(string $assetPath, string $packageName = null): string
+    {
+        if (null === $this->packages) {
+            throw new \Exception('To render the script or link tags, run "composer require symfony/asset".');
```

Comment:
All public APIs of this class are reaching this validation. So why is the Packages dependency an optional one in the constructor ?

Reply:
Good question. And, there may be a better solution. It's so that I can have:

```xml
<argument type="service" id="assets.packages" on-invalid="null" />
```

For this service's definition. I suppose I could remove this service definition entirely in a compiler pass if `assets.packages` is not present. However, this `TagRenderer` service is itself required by the Twig extension. So *it* would need `on-invalid="null"` and to have some code to throw an exception in the correct situation. I'm not sure if one solution is absolutely better than another, or if there's a general standard.

Reply:
or remove also the Twig extension when it does not have the needed dependency, which is what we do in symfony itself.